app/src/test/java/org/mozilla/fenix/components/metrics/MetricControllerTest.kt

```diff
@@ -0,0 +1,113 @@
+import org.junit.Before
+import org.junit.Test
+
+class MetricControllerTest {
+
+    @MockK(relaxUnitFun = true) private lateinit var debugLogger: Logger
+    @MockK(relaxUnitFun = true) private lateinit var dataService: MetricsService
+    @MockK(relaxUnitFun = true) private lateinit var marketingService: MetricsService
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this)
+
+        mockkObject(MetricController.Companion)
+        every { MetricController.telemetryEnabled } returns true
+        every { MetricController.debugLogger } returns debugLogger
+        every { dataService.type } returns MetricServiceType.Data
+        every { marketingService.type } returns MetricServiceType.Marketing
+    }
+
+    @After
+    fun teardown() {
+        unmockkObject(MetricController.Companion)
+    }
+
+    @Test
+    fun `debug metric controller emits logs`() {
+        every { MetricController.telemetryEnabled } returns false
+        val controller = MetricController.create(mockk(), mockk(), mockk())
+
+        controller.start(MetricServiceType.Data)
+        verify { debugLogger.debug("DebugMetricController: start") }
+
+        controller.stop(MetricServiceType.Data)
+        verify { debugLogger.debug("DebugMetricController: stop") }
+
+        controller.track(Event.OpenedAppFirstRun)
+        verify { debugLogger.debug("DebugMetricController: track event: ${Event.OpenedAppFirstRun}") }
+    }
+
+    @Test
+    fun `release metric controller starts and stops services`() {
+        var enabled = true
+        val controller = MetricController.create(
+            listOf(marketingService, dataService),
+            isDataTelemetryEnabled = { enabled },
+            isMarketingDataTelemetryEnabled = { enabled }
+        )
+
+        excludeRecords { marketingService.type }
+        excludeRecords { dataService.type }
+
+        controller.start(MetricServiceType.Data)
+        controller.start(MetricServiceType.Marketing)
+        enabled = false
+        controller.stop(MetricServiceType.Marketing)
+        controller.stop(MetricServiceType.Data)
+
+        verifySequence {
+            dataService.start()
+            marketingService.start()
+            marketingService.stop()
+            dataService.stop()
+        }
+    }
+
+    @Test
+    fun `tracking events should be sent to matching service`() {
+        val controller = MetricController.create(
+            listOf(dataService, marketingService),
+            isDataTelemetryEnabled = { true },
+            isMarketingDataTelemetryEnabled = { true }
+        )
+        every { dataService.shouldTrack(Event.TabMediaPause) } returns false
+        every { marketingService.shouldTrack(Event.TabMediaPause) } returns true
+
+        controller.start(MetricServiceType.Marketing)
+        controller.track(Event.TabMediaPause)
+        verify { marketingService.track(Event.TabMediaPause) }
+    }
+
+    @Test
+    fun `tracking terminates if other service is not enabled`() {
```

Comment:
Is this intended behaviour? Tracking is stopped if there are any disabled services in the chain, instead of going to the first available service. (See also `MetricController.track`)

Reply:
cc @boek do you know about this?

Reply:
Ah, I think we wanted a `return@forEach` there.